File: TorrentAdder/SYListComputersVC.m

//
//  SYListComputersVC.m
//  TorrentAdder
//
//  Created by Stan Chevallier on 24/09/2015.
//  Copyright © 2015 Syan. All rights reserved.
//

#import "SYListComputersVC.h"
#import "SYComputerCell.h"
#import <GBPing.h>
#import "SYComputerModel.h"
#import "SYNetworkManager.h"
#import "SYNetworkModel.h"
#import "SYPinger.h"
#import "SYBonjourClient.h"
#import "SYEditComputerVC.h"

@interface SYListComputersVC () <UITableViewDataSource, UITableViewDelegate, GBPingDelegate>

@property (strong, nonatomic) UIBarButtonItem *buttonClose;
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (weak, nonatomic) IBOutlet UIProgressView *progressView;
@property (strong, nonatomic) NSMutableArray <SYComputerModel *> *computers;
@property (strong, nonatomic) SYPinger *pinger;

@end

@implementation SYListComputersVC

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.buttonClose = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemStop target:self action:@selector(buttonCloseTap:)];
    [self.navigationItem setLeftBarButtonItem:self.buttonClose];
    
    self.computers = [NSMutableArray array];
    [self.tableView registerNib:[UINib nibWithNibName:[SYComputerCell className] bundle:nil]
         forCellReuseIdentifier:[SYComputerCell className]];
    [self.tableView setTableFooterView:[[UIView alloc] init]];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self startPinging];
    [self setTitle:@"Add a computer"];
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    [self setTitle:@""];
}

- (void)startPinging
{
    if (self.pinger)
        return;
    
    __weak SYListComputersVC *wSelf = self;
    
    [self.progressView setProgress:0];
    
    self.pinger = [SYPinger pingerWithNetworks:[SYNetworkManager myNetworks:YES]];
    [self.pinger pingNetworkWithProgressBlock:^(CGFloat progress) {
        [wSelf.progressView setProgress:progress animated:YES];
    } validIpFoundBlock:^(NSString *ip) {
        [wSelf addComputerWithIP:ip];
    } finishedBlock:^(BOOL finished) {
        [wSelf.progressView setProgress:1 animated:YES];
    }];
}

- (void)reload
{
    [self.tableView reloadData];
}

- (void)buttonCloseTap:(id)sender {
    [self.presentingViewController dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark - Data

- (void)addComputerWithIP:(NSString *)ip
{
    SYComputerModel *computer = [[SYComputerModel alloc] initWithName:nil andHost:ip];
    [self.computers addObject:computer];
    [self.computers sortUsingComparator:^NSComparisonResult(SYComputerModel * _Nonnull obj1, SYComputerModel * _Nonnull obj2) {
        return [obj1.host compare:obj2.host options:NSNumericSearch];
    }];
    
    [self.tableView beginUpdates];
    [self.tableView insertRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:[self.computers indexOfObject:computer] inSection:0]]
                          withRowAnimation:UITableViewRowAnimationAutomatic];
    [self.tableView endUpdates];
}

#pragma mark - UITableView

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.computers.count + 1;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    SYComputerCell *cell = [tableView dequeueReusableCellWithIdentifier:[SYComputerCell className]];
    
    if (indexPath.row >= self.computers.count)
        [cell setComputer:nil forAvailableComputersList:YES];
    else
        [cell setComputer:self.computers[indexPath.row] forAvailableComputersList:YES];
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    SYComputerModel *computer;
    
    if (indexPath.row >= self.computers.count)
        computer = [[SYComputerModel alloc] initWithName:nil andHost:nil];
    else
        computer = self.computers[indexPath.row];
    
    SYEditComputerVC *vc = [[SYEditComputerVC alloc] init];
    [vc setComputer:computer];
    [self.navigationController pushViewController:vc animated:YES];
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    return @"Available computers";
}

@end

File: TorrentAdder/SYListComputersVC.swift
//
//  SYListComputersVC.swift
//  TorrentAdder
//
//  Created by Stanislas Chevallier on 28/11/2018.
//  Copyright © 2018 Syan. All rights reserved.
//

import UIKit

class SYListComputersVC: UIViewController {

    // MARK: UIViewController
    override func viewDidLoad() {
        super.viewDidLoad()
        title = "Add a client"

        progressView.progressTintColor = .lightBlue
        
        let closeButton = UIBarButtonItem(barButtonSystemItem: .stop, target: self, action: #selector(self.closeButtonTap))
        navigationItem.leftBarButtonItem = closeButton
        
        tableView.registerCell(name: SYComputerCell.className)
        tableView.tableFooterView = UIView()
        
        NotificationCenter.default.addObserver(self, selector: #selector(self.hostnameResolverUpdatedNotification), name: .hostnameResolverUpdated, object: nil)
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self, name: .hostnameResolverUpdated, object: nil)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        startPinging()
    }
    
    // MARK: Properties
    private var availableIPs: [String] = []
    private var pinger: SYPinger?
    // MARK: View
    @IBOutlet private var tableView: UITableView!
    @IBOutlet private var progressView: UIProgressView!
    
    // MARK: Actions
    @objc private func closeButtonTap() {
        dismiss(animated: true, completion: nil)
    }
    
    // MARK: Notifications
    @objc private func hostnameResolverUpdatedNotification() {
        reload()
    }
    
    // MARK: Content
    private func reload() {
        tableView.reloadData()
    }
    
    private func addAvailableIP(_ ip: String?) {
        guard let ip = ip else { return }
        availableIPs.append(ip)
        availableIPs.sort { (ip1, ip2) -> Bool in
            return ip1.compare(ip2, options: .numeric) == .orderedAscending
        }
        
        if let index = availableIPs.index(of: ip) {
            tableView.beginUpdates()
            tableView.insertRows(at: [IndexPath(row: index, section: 0)], with: .automatic)
            tableView.endUpdates()
        }
        else {
            tableView.reloadData()
        }
    }
    
    // MARK: Ping
    private func startPinging() {
        guard pinger == nil else { return }
        
        pinger = SYPinger(networks: SYIPv4Interface.deviceNetworks(filterLocalInterfaces: true))
        pinger?.progressBlock = { [weak self] (progress) in
            self?.progressView.setProgress(Float(progress), animated: true)
        }
        pinger?.ipFoundBlock = { [weak self] (ip) in
            self?.addAvailableIP(ip)
        }
        pinger?.finishedBlock = { [weak self] (finished) in
            self?.progressView?.setProgress(1, animated: true)
        }
        
        progressView.progress = 0
        pinger?.start()
    }
}

extension SYListComputersVC : UITableViewDataSource {
    func client(at indexPath: IndexPath) -> SYClient? {
        guard indexPath.row < availableIPs.count else { return  nil }
        let host = availableIPs[indexPath.row]
        let name = SYHostnameResolver.shared.hostnameForIP(host) ?? host
        return SYClient(host: host, name: name)
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return availableIPs.count + 1
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueCell(type: SYComputerCell.self, indexPath: indexPath)
        cell.isDiscoveredClient = true
        cell.client = client(at: indexPath)
        return cell
    }
    
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return "Available clients"
    }
}

extension SYListComputersVC : UITableViewDelegate {
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        
        let vc = SYEditComputerVC()
        vc.client = client(at: indexPath) ?? SYClient(host: "", name: nil)
        navigationController?.pushViewController(vc, animated: true)
    }
}
